### ISSUE TITLE
moe_kernels.py, stage-2 reduce path: small SGLANG_MORI_NUM_MAX_DISPATCH_TOKENS_PER_RANK silently corrupts output (gsm8k=0) at low concurrency

### ISSUE THREAD

**Issue by Oseltamivir** [sgl-project/sglang#27194]
## Summary

On the MoRI expert-parallel backend, a small `SGLANG_MORI_NUM_MAX_DISPATCH_TOKENS_PER_RANK` **silently corrupts decode output** — the model still generates fluent-looking tokens (speculative acceptance length stays high), but the content is semantically wrong and gsm8k drops to **0.00**. The value is forwarded to MoRI's `max_num_inp_token_per_rank` with no minimum/validation, so configs that legitimately produce a small per-rank dispatch buffer (low concurrency × dp-attention × MTP) hit it with no error or warning.

This is a silent correctness bug: nothing crashes, throughput even looks marginally *better*, so it is easy to ship a broken Pareto point.

## Environment

- Image: `lmsysorg/sglang-rocm:v0.5.12.post1-rocm720-mi35x` (sglang v0.5.12.post1)
- HW: AMD MI355X (CDNA, warpSize 64)
- Model: DeepSeek-R1 (FP4), prefill/decode disaggregated
- Decode parallelism: TP8 / EP8 / **dp-attention** / **MTP (num-draft-tokens=4)**
- Backends: `--moe-a2a-backend mori --disaggregation-transfer-backend mori --deepep-mode normal`

## Symptom

At decode concurrency 64 with TP8 + MTP3, the per-rank dispatch buffer works out to `64/8 * 4 = 32` tokens. The server starts, warms up, benchmarks at full speed, and produces grammatical output — but gsm8k = 0. Raising the buffer fixes it.

## Reproduction / evidence (shmoo of the env var, same config)

| `SGLANG_MORI_NUM_MAX_DISPATCH_TOKENS_PER_RANK` | gsm8k (strict) | output tok/s | run |
|---|---|---|---|
| 32 | **0.0000** | 6312 | https://github.com/SemiAnalysisAI/InferenceX/actions/runs/26913235190 |
| 64 (= warpSize) | **0.0000** | 6324 | https://github.com/SemiAnalysisAI/InferenceX/actions/runs/26919517564 |
| 256 | **0.9424** | 6219 | https://github.com/SemiAnalysisAI/InferenceX/actions/runs/26920499507 |

(Run links are internal CI; the relevant log excerpts are inline below.) All three runs generate identical token volume — 640 requests, 589,927 generated tokens — so the numbers are directly comparable. Note the corrupt runs are ~1.5% *faster* (less real work), which is why this slips through perf-only validation.

Eval log (256, correct):
```
[MoRI floor] DISPATCH_TOKENS=32 < floor 256; clamping to 256
Decode env: SGLANG_MORI_NUM_MAX_DISPATCH_TOKENS_PER_RANK=256
PASS: gsm8k exact_match,strict-match = 0.9424 (>= 0.85)
```

## This is not the LL vs non-LL kernel switch

`get_ep_dispatch_configs()` selects `InterNodeV1LL` vs `InterNodeV1` by comparing the dispatch size to `SGLANG_MORI_DISPATCH_INTER_KERNEL_SWITCH_THRESHOLD` (default 256). In the failing run that threshold was set to 16, and the decode log confirms:

```
[DP+EP override] Decode: ... DISPATCH_TOKENS=8, ... INTER_KERNEL_SWITCH=16
```

Since `32 > 16`, the **non-LL `InterNodeV1`** kernel is selected — and it still corrupts. So the bug is the per-rank buffer size itself, not the kernel variant.

## Where it originates in sglang

In `python/sglang/srt/layers/moe/token_dispatcher/moriep.py` (v0.5.12):

- The value is read with a default of 4096 and **no lower bound**:
  ```python
  self.num_max_dispatch_tokens_per_rank = get_int_env_var(
      "SGLANG_MORI_NUM_MAX_DISPATCH_TOKENS_PER_RANK", 4096)
  ```
- It is passed straight through to MoRI as `max_num_inp_token_per_rank=num_max_dispatch_tokens_per_rank`.
- There is already a `if num_max_dispatch_tokens_per_rank < 128:` branch that swaps the launch config (`warp_num_per_block`), so small dispatch sizes are clearly known to be special — but that branch does **not** prevent the corruption at 32/64.

## ~~Likely mechanism (hypothesis — root cause is in MoRI)~~ — **CORRECTED, see below**

~~The MoRI dispatch kernel writes received tokens into warpSize-aligned slots: `destTokId = flagSlotId * warpSize + laneId`, into a per-PE region whose stride is `max_num_inp_token_per_rank`. When that region is smaller than warpSize (64), or smaller than the routing fan-in a rank can receive from all peers, the writes go out of bounds. Empirically warpSize (64) is necessary but not sufficient; ≥256 is needed for this config. Filed in parallel against ROCm/mori: `src/ops/dispatch_combine/internode_v1.cpp` (`DispatchInterNodeSend` / `DispatchIntraNodeBlock`). https://github.com/ROCm/mori/issues/356~~

> ### UPDATE, corrected root cause (not MoRI issue)
>
> The MoRI hypothesis is wrong, and `ROCm/mori#356` was closed as not-a-MoRI-issue. No out-of-bounds write in the dispatch kernel is involved.
>
> The actual root cause is prob in AITER's FlyDSL MoE stage-2 reduce path (`aiter/ops/flydsl/moe_kernels.py`), confirmed by @billishyahao:
>
> - The reduce intermediate is allocated with `torch.empty` (**uninitialized**).
> - In EP, gemm2 writes only valid local-expert slots; masked (non-local) slots are never written.
> - `torch.sum` over `topk` then folds those uninitialized slots in → NaN/garbage → corrupt-but-fluent output.
> - Only the **M≈1024 tuning bucket** selects this reduce kernel; other buckets use a zero-init atomic accumulate, which is why M=1032/2048 were fine and only specific low-concurrency points corrupt.
>
> **Fixed in [ROCm/aiter#3377](https://github.com/ROCm/aiter/pull/3377)** (masked gather `valid = expert_mask[topk_ids] != 0` + EP tuned-key `topk -= is_ep`), pulled into SGLang via **#27555** (AITER commit bump `6740fb7` → `7a8ff7d`).
>
> **Not fully resolved:** with the fix applied, conc64 passes but **conc32 and conc16 still fail** ([run](https://github.com/SemiAnalysisAI/InferenceX/actions/runs/27225030090)) — the sub-1024 M-buckets likely need the same masked-reduce treatment. Keeping this open until those are covered.

## Requested fix

1. In `moriep.py`, **enforce or validate a minimum** for `num_max_dispatch_tokens_per_rank` (and/or emit a clear warning) instead of silently forwarding a too-small value.
2. Revisit the `< 128` launch-config branch — it does not make small buffers correct.
3. Coordinate the actual kernel-level fix with ROCm/mori (sizing from fan-in, or asserting a minimum).

At minimum, a silent correctness failure should become a loud one.

**Comment by functionstackx** [comment]
@billishyahao @HaiShaw @Duyi-Wang @chunfangamd @andyluo7

**Comment by functionstackx** [comment]
after fixing this, Can we please add tests & assertions in sglang and/or MoRI to ensure that this error doesn't siliently fail again?

**Comment by billishyahao** [comment]
Hi @functionstackx @Oseltamivir  

This issue can be reproduced on recent nightly build as well: `lmsysorg/sglang-rocm:v0.5.12.post1-rocm720-mi35x-20260529`

with the following command:

```bash
SGLANG_USE_AITER=1                      \
SGLANG_MORI_DISPATCH_DTYPE=bf16         \
SGLANG_MORI_COMBINE_DTYPE=bf16               \
SGLANG_MORI_QP_PER_TRANSFER=4           \
SGLANG_MORI_NUM_WORKERS=4               \
MORI_IO_SQ_BACKOFF_TIMEOUT_US=50000    \
MORI_IO_QP_MAX_SEND_WR=16384           \
MORI_IO_QP_MAX_CQE=32768               \
MORI_IO_QP_MAX_SGE=4                   \
SGLANG_DISAGGREGATION_BOOTSTRAP_TIMEOUT=3600  \
SGLANG_DISAGGREGATION_WAITING_TIMEOUT=3600    \
MORI_SHMEM_MODE=ISOLATION                     \
SGLANG_ENABLE_SPEC_V2=1                       \
SGLANG_ENABLE_OVERLAP_PLAN_STREAM=0           \
SGLANG_LOG_MS=true                            \
SGLANG_DISAGGREGATION_NUM_PRE_ALLOCATE_REQS=32 \
MORI_EP_LAUNCH_CONFIG_MODE=AUTO                         \
MORI_APP_LOG_LEVEL=INFO  \
MORI_RDMA_SL=3   \
MORI_RDMA_TC=96  \
MORI_IO_SL=3     \
MORI_IO_TC=96    \
MORI_IO_TC_DISABLE=0  \
NCCL_IB_HCA=ionic_0,ionic_1,ionic_2,ionic_3,ionic_4,ionic_5,ionic_6,ionic_7 \
GLOO_SOCKET_IFNAME=enp81s0f1 \
NCCL_SOCKET_IFNAME=enp81s0f1 \
SGLANG_MORI_NUM_MAX_DISPATCH_TOKENS_PER_RANK=128 \
python3 -m sglang.launch_server --model-path /models/DeepSeek-R1-0528-MXFP4-v2 --disaggregation-mode decode --disaggregation-ib-device ionic_0,ionic_1,ionic_2,ionic_3,ionic_4,ionic_5,ionic_6,ionic_7 --host 0.0.0.0 --port 8321 --trust-remote-code --tp-size 8 --ep-size 8 --dp-size 8 --decode-log-interval 1000 --log-level info --watchdog-timeout 3600 --load-balance-method round_robin --kv-cache-dtype fp8_e4m3 --attention-backend aiter --disaggregation-transfer-backend mori --moe-a2a-backend mori --deepep-mode normal --enable-dp-attention --moe-dense-tp-size 1 --enable-dp-lm-head  --mem-fraction-static 0.85 --max-running-requests 8 --disable-cuda-graph --prefill-round-robin-balance
```

The issue is gone after applying env SGLANG_MORI_NUM_MAX_DISPATCH_TOKENS_PER_RANK=129

The root cause is narrowed down to flydsl moe `aiter/ops/flydsl/moe_kernels.py`, stage-2 reduce path:
```bash
target = torch.empty((token_num * topk * model_dim,), ...)   # uninitialized
_run_compiled(exe, args)                                      # gemm2 writes ONLY valid sorted slots
torch.sum(target.view(token_num, topk, model_dim), dim=1, out=out)
```
- The reduce intermediate is `torch.empty` (uninitialized).
- gemm2 writes only valid token-expert slots; in EP, slots for non-local (masked) experts are never written.
- torch.sum over topk adds those uninitialized slots → sometimes NaN/garbage → non-deterministic.

Only the M≈1024 tuning bucket picks the reduce kernel; other buckets use atomic (zero-init, valid-only), so M=1032/2048 are fine. A NaN in any of
~60 layers × decode steps propagates to garbled output.

Fixed it through torch.empty → torch.zeros for the reduce intermediate (masked slots contribute 0). Equivalent in intent to https://github.com/ROCm/aiter/pull/3377, which instead skips invalid slots via expert_mask[topk_ids] in the reduction kernel.

This issue have been resolved in https://github.com/ROCm/aiter/pull/3377 so please wait for the sglang side to bump the aiter version to obtain https://github.com/ROCm/aiter/pull/3377

cc @HaiShaw @Duyi-Wang

**Comment by functionstackx** [comment]
hi @billishyahao thanks for the debug & confirming the AMD AITER bug! when should we expect SGLang to update to the latest AITER version that includes this fix and when should we expect when u will update the InferenceX PR with an accuracy fix? 

this bug effects all of ur recent InferenceX submissions for deepseek r1 DI FP4 MTP for conc64,32,16,8, 4,2, etc

**Comment by functionstackx** [comment]
furthermore, can AMD SGLang tests be will be added that would prevent accuracy issues like this due to downstream AMD AITER dependencies?

**Comment by chunfangamd** [comment]
cc: @functionstackx 

https://github.com/sgl-project/sglang/pull/27555

SGLang： Upgrade the AITER commit pinned in the ROCm Docker image to pick up the latest AITER fixes/improvements.
This bumps AITER_COMMIT_DEFAULT from 6740fb7911db231e6eadad45194ad6713d25f9fb to 7a8ff7dd4ae3063ff1a18622a46460125c84370e

**Comment by functionstackx** [comment]
hi @chunfangamd @billishyahao  @sunway513 @HaiShaw @yadaish thanks for the inital debug & and confirming the bug & the inital fix attempt. But unfortunately even have https://github.com/sgl-project/sglang/pull/27555   https://github.com/ROCm/aiter/pull/3377 , conc64 passes but conc32 & conc16 fail. Can u take a look into this? Attached below is the link with the logs 

https://github.com/SemiAnalysisAI/InferenceX/actions/runs/27225030090

**Comment by HaiShaw** [comment]
@functionstackx, we are checking.

**Comment by functionstackx** [comment]
maybe if there is enough nodes, implementing multinode nightly DI CIs would help prevent these accuracy issues like what CUDA side has https://github.com/sgl-project/sglang/issues/22464

**Comment by HaiShaw** [comment]
@functionstackx - Sorry for a bit delay in response:

After upgrading AITER, the original issue reported here, where GSM8K accuracy dropped to 0, has been resolved.
However, we observed a different issue only at DSr1 with concurrency = 64, where GSM8K accuracy drops from 90%+ to 80%+.
 
During our investigation, we found that:
- Accuracy remains stable at around 95% under other concurrency levels.
- The accuracy degradation at concurrency 64 can be mitigated by reducing quantization aggressiveness, such as:
  - Disabling FP8 KV cache
  - Using a less aggressive quantization configuration than Quark MXFP4
 
These findings suggest that the issue is likely related to kernels exercised specifically under the concurrency-64 workload. The root cause may be numerical accuracy degradation or improper handling of certain corner cases in those kernels, rather than a functional correctness issue.
 
Given that this behavior has only been observed on DeepSeek-R1 so far, further root-cause analysis of such a subtle accuracy regression would require some effort. Considering the limited impact and lack of reproduction on other models, we recommend closing this issue for now.
 
If similar behavior is observed elsewhere, we can reopen the issue and continue the investigation. Since the original issue reported here (GSM8K accuracy dropping to 0) has already gone by the AITER upgrade, we believe this original issue can be closed.

cc @Duyi-Wang

**Comment by functionstackx** [comment]
hi @HaiShaw thanks for the detailed write up.

I think this is still quite an serious issue that conc64 is at 80% on simple grade school math GSM8k task verus the expected behavior of 94%+.

Understood the priorities of models, though not sure the reasoning why AMD submitted DeepSeek R1 in early May 2026 but in late May 2026, it isn't an priority anymore.

I think if AMD does have the bandwidth to fix the 80% accuracy we should still leave this issue open such that other people that run across this issue or similar kernel accuracy can see that this is an known issue and not planned for fixing.

**Comment by Duyi-Wang** [comment]
@Oseltamivir  @functionstackx  That accuracy drop issue in conc64 has been fixed, it's caused by read-after-write synchronization race in aiter persist mla kernel. Detail analysis in https://github.com/ROCm/aiter/pull/4420. This fixed in aiter is included in https://github.com/ROCm/aiter/commit/cdd6628f9 and SGLang upstream image contains this. Below is verification ci running:
- FAIELD 0727 image + SGLANG_AITER_MLA_PERSIST=1 with old CI script  https://github.com/SemiAnalysisAI/InferenceX/actions/runs/30363175172
- PASSED 0727 image + fix patch,  SGLANG_AITER_MLA_PERSIST=1 with old CI script https://github.com/SemiAnalysisAI/InferenceX/actions/runs/30363185458

Latest CI running failed in InferenceX is due to SGLANG_AITER_MLA_PERSIST has been set to 0 in CI scripts, like this CI https://github.com/SemiAnalysisAI/InferenceX/actions/runs/30604299580/attempts/1?pr=2435
This is another issue, gsm8k score dropped to 0 when conc > 1024 when SGLANG_AITER_MLA_PERSIST=0. This issue has been located and fixed, fix patch and detail analysis are in the https://github.com/ROCm/aiter/pull/4587
